File: leetcode/cpp/150. 逆波兰表达式求值.cpp
class Solution {
public:
    int add(int x, int y) {return x + y;}
    int sub(int x, int y) {return x - y;}
    int mul(int x, int y) {return x * y;}
    int div(int x, int y) {return x / y;}
    typedef int (Solution::*func)(int, int);

    int evalRPN(vector<string>& tokens) {
        int x, y;
        stack<int> st;
        unordered_map<string, func> d;
        d["+"] = &Solution::add;
        d["-"] = &Solution::sub;
        d["*"] = &Solution::mul;
        d["/"] = &Solution::div;
        for (const string& curr : tokens) {
            if (d.count(curr) > 0) {
                y = st.top(), st.pop();
                x = st.top(), st.pop();
                st.push((this->*d[curr])(x, y));
            } else {
                st.push(stoi(curr));
            }
        }
        return st.top();
    }
};

File: leetcode/python/150. 逆波兰表达式求值.py
#!/bin/bash python
"""
150. 逆波兰表达式求值

思路：
栈
"""

class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        """
        :type tokens: List[str]
        :rtype: int
        应该就是直接栈操作了
        """
        stack = []
        d = {"+": lambda x, y : x + y,
             "-": lambda x, y : x - y,
             "*": lambda x, y : x * y,
             "/": lambda x, y : int(x / y)}
        for char in tokens:
            if char in d:
                y = stack.pop()
                x = stack.pop()
                stack.append(d[char](x, y))
            else:
                stack.append(int(char))
        return stack[-1]